mainchain/api_interface.go

```diff
@@ -33,28 +37,39 @@ import (
 
 type APIBackend interface {
 	// Blockchain API
-	HeaderByNumber(ctx context.Context, number rpc.BlockNumber) *types.Header
-	HeaderByHash(ctx context.Context, hash common.Hash) *types.Header
-	HeaderByNumberOrHash(ctx context.Context, blockNrOrHash rpc.BlockNumberOrHash) (*types.Header, error)
 	BlockByNumber(ctx context.Context, number rpc.BlockNumber) *types.Block
 	BlockByHash(ctx context.Context, hash common.Hash) *types.Block
 	BlockByNumberOrHash(ctx context.Context, blockNrOrHash rpc.BlockNumberOrHash) (*types.Block, error)
 	BlockInfoByBlockHash(ctx context.Context, hash common.Hash) *types.BlockInfo
-	StateAndHeaderByNumber(ctx context.Context, number rpc.BlockNumber) (*state.StateDB, *types.Header, error)
-	StateAndHeaderByNumberOrHash(ctx context.Context, blockNrOrHash rpc.BlockNumberOrHash) (*state.StateDB, *types.Header, error)
+
 	GetKVM(ctx context.Context, msg types.Message, state *state.StateDB, header *types.Header) (*kvm.KVM, func() error, error)
 	GetValidators() ([]*staking.Validator, error)
 	GetValidator(valAddr common.Address) (*staking.Validator, error)
 	GetValidatorCommission(valAddr common.Address) (uint64, error)
 	GetDelegationsByValidator(valAddr common.Address) ([]*staking.Delegator, error)
+
+	HeaderByNumber(ctx context.Context, number rpc.BlockNumber) *types.Header
```

Comment:
Consider remove this since we have HeaderByNumberOrHash

Reply:
We need both in core APIs and filter APIs. `HeaderByNumberOrHash` method may confuse caller and wrong caching if it's called with both `hash` and `height`. For example:

```go
// GetHeader retrieves a block header from the database by hash and height,
// caching it if found.
func (hc *HeaderChain) GetHeader(hash common.Hash, height uint64) *types.Header {
	// Short circuit if the header's already in the cache, retrieve otherwise
	if header, ok := hc.headerCache.Get(hash); ok {
		return header.(*types.Header)
	}
	header := hc.kaiDb.ReadHeader(height)
	if header == nil {
		return nil
	}
	// Cache the found header for next time and return
	hc.headerCache.Add(hash, header)
	return header
}
```

If we provide a `height = 100` and the empty block hash `common.Hash{}`, the node will query the header of block `100` and cache it with `key = common.Hash{}, value = header of block 100`. The next `GetHeader` calls will return the wrong cached block header if the caller provide the empty block hash `common.Hash{}`, with any `height`.